src/main/java/org/opensearch/security/resources/migrate/MigrateResourceSharingInfoApiAction.java

```diff
@@ -0,0 +1,309 @@
+ */
+public class MigrateResourceSharingInfoApiAction extends AbstractApiAction {
+
+    private static final Logger LOGGER = LogManager.getLogger(MigrateResourceSharingInfoApiAction.class);
+
+    private static final List<Route> routes = addRoutesPrefix(ImmutableList.of(new Route(POST, "/resources/migrate")));
+
+    private final ResourceSharingIndexHandler sharingIndexHandler;
+
+    public MigrateResourceSharingInfoApiAction(
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        SecurityApiDependencies securityApiDependencies,
+        ResourceSharingIndexHandler sharingIndexHandler
+    ) {
+        super(Endpoint.RESOURCE_SHARING, clusterService, threadPool, securityApiDependencies);
+        this.sharingIndexHandler = sharingIndexHandler;
+        this.requestHandlersBuilder.configureRequestHandlers(this::migrateApiRequestHandlers);
+    }
+
+    @Override
+    public List<Route> routes() {
+        return routes;
+    }
+
+    @Override
+    protected CType<?> getConfigType() {
+        return null;
+    }
+
+    private void migrateApiRequestHandlers(RequestHandler.RequestHandlersBuilder b) {
+        b.allMethodsNotImplemented().override(POST, this::handleMigrate);
+    }
+
+    private void handleMigrate(RestChannel channel, RestRequest request, Client client) throws IOException {
+        endpointValidator.createRequestContentValidator()
+            .validate(request)
+            // 1. read the old sharing docs
+            .map(pair -> loadCurrentSharingInfo(request, client))
+            // 2. transform and index the new ResourceSharing records
+            .map(this::createNewSharingRecords)
+            // 3. send the response
+            .valid(summary -> { ok(channel, summary); })
+            .error((status, toXContent) -> response(channel, status, toXContent));
+    }
+
+    /**
+     * This method:
+     *      1. Pulls "source_index" from the request body
+     *      2. Does a match_all search (up to 10k) in the "source_index"
+     *      3. Create a SourceDoc for each raw doc
+     *      3. Returns a triple of the source index name, the default access level and the list of source docs.
+     */
+    private ValidationResult<Triple<String, String, List<SourceDoc>>> loadCurrentSharingInfo(RestRequest request, Client client)
+        throws IOException {
+        JsonNode body = Utils.toJsonNode(request.content().utf8ToString());
+
+        String sourceIndex = body.get("source_index").asText();
+        String userNamePath = body.get("user_name_path").asText();
+        String backendRolesPath = body.get("backend_roles_path").asText();
+        String defaultAccessLevel = body.has("default_access_level") ? body.get("default_access_level").asText() : PLACE_HOLDER;
+
+        SearchRequest sr = new SearchRequest(sourceIndex);
+        sr.source(new SearchSourceBuilder().query(QueryBuilders.matchAllQuery()).size(10_000)); // TODO verify
```

Comment:
why not use scroll to ensure all records are traversed?

Reply:
done